File: game/bounding-box.go
package game

type BoundingBoxCorners struct {
	Top    float64 `json:"top"`
	Right  float64 `json:"right"`
	Bottom float64 `json:"bottom"`
	Left   float64 `json:"left"`
}

func NewBoundingBoxCorners(top float64, right float64, bottom float64, left float64) BoundingBoxCorners {
	return BoundingBoxCorners{top, right, bottom, left}
}

type BoundingBox struct {
	Position Point              `json:"position"`
	Size     Size               `json:"size"`
	Corners  BoundingBoxCorners `json:"corners"`
}

func NewBoundingBox(size Size, position Point) BoundingBox {
	bb := BoundingBox{
		Size:     size,
		Position: position,
		Corners:  NewBoundingBoxCorners(0.0, 0.0, 0.0, 0.0),
	}
	return bb
}

func (bb *BoundingBox) Update() {
	bb.Corners.Top = bb.Position.Y - bb.Size.Height/2.0
	bb.Corners.Right = bb.Position.X + bb.Size.Width/2.0
	bb.Corners.Bottom = bb.Position.Y + bb.Size.Height/2.0
	bb.Corners.Left = bb.Position.X - bb.Size.Width/2.0
}

func (bb *BoundingBox) SetSize(width float64, height float64) {
	bb.Size.Width = width
	bb.Size.Height = height
	bb.Update()
}

func (bb *BoundingBox) SetPosition(position Point) {
	bb.Position = position
	bb.Update()
}

func (bb *BoundingBox) IsPointInside(point Point) bool {
	return CollisionWithPoint(point, bb.Corners)
}

File: src-tauri/src/game/bounding_box.rs
use serde::{Deserialize, Serialize};
use ts_rs::TS;

use crate::game::point::Point;
use crate::game::size::Size;

use super::collision::{collision_with_bb, collision_with_point};

#[derive(Serialize, Deserialize, Debug, Clone, TS)]
#[ts(export)]
pub struct BoundingBoxEdges {
    pub top: f32,
    pub right: f32,
    pub bottom: f32,
    pub left: f32,
}

impl BoundingBoxEdges {
    pub fn new(top: f32, right: f32, bottom: f32, left: f32) -> BoundingBoxEdges {
        BoundingBoxEdges {
            top,
            right,
            bottom,
            left,
        }
    }

    pub fn flatten(&mut self) -> (f32, f32, f32, f32) {
        (self.top, self.right, self.bottom, self.left)
    }
}

#[derive(Serialize, Deserialize, Debug, Clone, TS)]
#[ts(export)]
pub struct BoundingBox {
    position: Point,
    size: Size,
    edges: BoundingBoxEdges,
}

impl BoundingBox {
    pub fn new(size: Size, position: Point) -> BoundingBox {
        let mut bb = BoundingBox {
            size,
            position,
            edges: BoundingBoxEdges::new(0.0, 0.0, 0.0, 0.0),
        };
        bb.update();
        bb
    }

    fn update(&mut self) {
        self.edges.top = self.position.y - self.size.height / 2.0;
        self.edges.right = self.position.x + self.size.width / 2.0;
        self.edges.bottom = self.position.y + self.size.height / 2.0;
        self.edges.left = self.position.x - self.size.width / 2.0;
    }

    pub fn set_size(&mut self, size: Size) {
        self.size = size;
        self.update();
    }

    pub fn get_size(&mut self) -> Size {
        self.size.clone()
    }

    pub fn set_position(&mut self, position: Point) {
        self.position = position;
        self.update();
    }

    pub fn get_position(&mut self) -> Point {
        self.position.clone()
    }

    pub fn point_inside(&mut self, point: (f32, f32)) -> bool {
        collision_with_point((self.edges.top, self.edges.right, self.edges.bottom, self.edges.left), point)
    }

    pub fn does_intersects(&mut self, bb: (f32, f32, f32, f32)) -> bool {
        collision_with_bb((self.edges.top, self.edges.right, self.edges.bottom, self.edges.left), bb)
    }

    pub fn does_intersects_bb(&mut self, bb: BoundingBox) -> bool {
        self.point_inside((bb.edges.left, bb.edges.top))
            || self.point_inside((bb.edges.right, bb.edges.top))
            || self.point_inside((bb.edges.right, bb.edges.bottom))
            || self.point_inside((bb.edges.left, bb.edges.bottom))
    }
}
